### DISCUSSION TITLE
User profile attributes vs. user attributes

### DISCUSSION THREAD

**Discussion by DaBonse** [keycloak/keycloak#28943] [Q&A] [answered]
Hello,

I ask myself what is the different between the new feature user profile attributes and the user attributes in the old user representation.

The old user attributes is in my opinion a container to save custom meta data to the user context, e.g. userAcceptCookie => true. Other data like email or username are not an attribute.

See also user representation in the REST API, for a better understanding:
https://www.keycloak.org/docs-api/latest/rest-api/index.html#UserRepresentation

But in the new user profile, the documentation speaks from user attributes. Predefined attributes are username, email, firstName and lastName.
These attributes can be managed or unmanaged. Unmanaged attributes are not defined user attributes, but existing elements under the user attribute section.

Now I came to my questions:

1.) Is an user profile attribute technically the same like the old user attribute? So I can manage user attributes over REST API in this way?
2.) The four predefined attributes like username, will these attributes saved under the attribute section? And will these elements removed in future from the user representation.

Thanks,

DaBonse

**Comment by pedroigor** [comment, marked as answer]
> 1.) Is an user profile attribute technically the same like the old user attribute? So I can manage user attributes over REST API in this way?

Yes, they are technically the same and you can manage them over REST just like before. However, the declarative user profile (now enabled by default) introduces conditions on how these attributes are managed so that (by default) only the attributes explicitly set in your user profile configuration are "manageable". The reason for being more strict about attributes is backed by security concerns and by UX improvements to dynamically render/process custom attributes without forcing you to customize the server templates, authenticators, or required actions.

Any other attribute is an "unmanaged" attribute and is ignored by the server if not available in the user profile configuration. For these attributes, we have those `Unmanaged Attribute` policies (ENABLED, ADMIN_VIEW, ADMIN_EDIT) to allow you to relax how these attributes are managed. For instance, the `ENABLED` policy is pretty much the old behavior where any attribute is accepted/recognized by the server and it is only there for migration purposes.

> 2.) The four predefined attributes like username, will these attributes saved under the attribute section? And will these elements removed in future from the user representation.

Even before the user profile feature, these attributes were managed as regular attributes but persisted as a field/column in the model/table. So yeah, they are also manageable as a regular attribute in the attributes map but they are not persisted as a regular attribute.

Behind the scenes, when you set those "root" attributes to the representation/model they are automatically added to the attribute map and vice-versa.

I'm not sure about removing those "root" attributes from the representation but it is technically possible to do that and leverage the attributes map to manage them.

**Reply by DaBonse** [reply]
Thanks a lot pedroigor.

I think, user profile is the right way.
